### PR TITLE
pkg/fileutil: skip TestIsDirWriteable when running as root

### LINKED ISSUE #4007: TestIsDirWriteable fails with root privilege
```
--- FAIL: TestIsDirWriteable (0.00s)
        fileutil_test.go:38: expected IsDirWriteable to error
```

The bug happens on a907ca5e62503e9ee2f3db95024f00e92e374586.
I will submit a patch later.

### PR DESCRIPTION
**Pull request by AkihiroSuda**
Closes #4007

### CODE REVIEW

File: pkg/fileutil/fileutil_test.go [hunk 1]
```diff
@@ -34,8 +35,14 @@ func TestIsDirWriteable(t *testing.T) {
 	if err := os.Chmod(tmpdir, 0444); err != nil {
 		t.Fatalf("unexpected os.Chmod error: %v", err)
 	}
-	if err := IsDirWriteable(tmpdir); err == nil {
-		t.Fatalf("expected IsDirWriteable to error")
+	me, err := user.Current()
```
Comment (xiang90) [inline pkg/fileutil/fileutil_test.go]: let's just fatal the test if cannot get user. 

```
if err != nil {
    t.Fatal("failed to get current user")
}
```

File: pkg/fileutil/fileutil_test.go [hunk 2]
```diff
@@ -34,8 +35,14 @@ func TestIsDirWriteable(t *testing.T) {
...
+	if err == nil {
+		if ! (me.Name == "root" || me.Name == "Administrator") {
+			// ideally we should check CAP_DAC_OVERRIDE. but it does not matter for tests.
+			if err := IsDirWriteable(tmpdir); err == nil {
+				t.Fatalf("expected IsDirWriteable to error")
```
Comment (xiang90) [inline pkg/fileutil/fileutil_test.go]: let's change t.Fatalf to t.Errorf since this is a test failure

**Comment by xiang90** [comment]
LGTM besides the nit. 

@heyitsanthony Can you please take a look too?

**Comment by xiang90** [comment]
Also we need to gofmt this.

File: pkg/fileutil/fileutil_test.go (hunk 1 above)
Comment (AkihiroSuda) [inline pkg/fileutil/fileutil_test.go]: err can be non-nil when cross-compiled.
So I would like to keep current behavior.
https://www.google.com/search?q=user%3A+Current+not+implemented

Comment (xiang90) [inline pkg/fileutil/fileutil_test.go]: make sense.

Comment (xiang90) [inline pkg/fileutil/fileutil_test.go]: let add a comment for this behavior.

**Comment by AkihiroSuda** [comment]
@xiang90 Thank you for the reviewing.
I updated the PR. (53a96a1)

**Comment by gyuho** [comment]
`govet -shadow` failed at tip but not critical. https://travis-ci.org/coreos/etcd/jobs/98263586#L296-L298

I can easily fix after this merge.

LGTM.

File: pkg/fileutil/fileutil_test.go [hunk 3]
```diff
@@ -34,8 +35,19 @@ func TestIsDirWriteable(t *testing.T) {
 	if err := os.Chmod(tmpdir, 0444); err != nil {
 		t.Fatalf("unexpected os.Chmod error: %v", err)
 	}
-	if err := IsDirWriteable(tmpdir); err == nil {
-		t.Fatalf("expected IsDirWriteable to error")
+	me, err := user.Current()
+	if err != nil {
+		// err can be non-nil when cross compiled
+		// http://stackoverflow.com/questions/20609415/cross-compiling-user-current-not-implemented-on-linux-amd64
+		t.Logf("failed to get current user: %v", err)
```
Comment (jonboulle) [inline pkg/fileutil/fileutil_test.go]: Why not Skipf?

File: pkg/fileutil/fileutil_test.go [hunk 4]
```diff
@@ -34,8 +35,19 @@ func TestIsDirWriteable(t *testing.T) {
 	if err := os.Chmod(tmpdir, 0444); err != nil {
 		t.Fatalf("unexpected os.Chmod error: %v", err)
 	}
-	if err := IsDirWriteable(tmpdir); err == nil {
-		t.Fatalf("expected IsDirWriteable to error")
+	me, err := user.Current()
+	if err != nil {
```
Comment (jonboulle) [inline pkg/fileutil/fileutil_test.go]: how about
if err != nil: t.Skipf("failed to get ..")
if me.Name == root, administrator: t.Skipf("running as superuser ..")

**Comment by heyitsanthony** [comment]
lgtm

File: pkg/fileutil/fileutil_test.go (hunk 4 above)
Comment (AkihiroSuda) [inline pkg/fileutil/fileutil_test.go]: @jonboulle Thanks, updated PR.

**Comment by xiang90** [comment]
@AkihiroSuda Thanks for the fix! Merging.

**Pull request merged**